File: build_db_sql.js
const fs = require('fs');
const path = require('path');
const text = fs.readFileSync(path.join(__dirname, 'app.js'), 'utf8');
const match = text.match(/const DB = \{([\s\S]*?)\};\s*\/\/ ══ STATE/);
if (!match) throw new Error('DB object not found');
const body = match[1];
const obj = eval('({' + body + '})');
console.log('tables', Object.keys(obj));
for (const [k, v] of Object.entries(obj)) {
  console.log('# TABLE', k, v.length, 'rows');
  if (v.length === 0) continue;
  const fields = Array.from(new Set(v.flatMap(r => Object.keys(r)))).sort();
  for (const row of v) {
    const vals = fields.map(f => {
      const val = row[f];
      if (val === undefined || val === null) return 'NULL';
      if (typeof val === 'number') return String(val);
      return "'" + String(val).replace(/'/g, "''") + "'";
    });
    const cols = fields.map(f => '`' + f + '`').join(',');
    console.log('INSERT INTO `' + k + '` (' + cols + ') VALUES (' + vals.join(',') + ');');
  }
}

File: build_db_sql.py
import re
import ast
from pathlib import Path

path = Path('app.js')
text = path.read_text(encoding='utf-8')

m = re.search(r'const DB = \{(.*?)\};\s*// ══ STATE', text, re.S)
if not m:
    raise SystemExit('DB object not found')
body = m.group(1)

# Add quotes to field names
body = re.sub(r'([\{,\s])(\w+)\s*:', r'\1"\2":', body)
# Convert JS strings to JSON-like strings
body = re.sub(r"'(\\.|[^'\\])*'", lambda x: '"' + x.group(0)[1:-1].replace('"', '\\"').replace('\\\\', '\\\\\\') + '"', body)
# Remove trailing commas
body = re.sub(r',\s*([\]}])', r'\1', body)
json_text = '{' + body + '}'
json_text = json_text.replace('null', 'None').replace('true', 'True').replace('false', 'False')
obj = ast.literal_eval(json_text)

print('tables', list(obj.keys()))
for k, v in obj.items():
    print('# TABLE', k, len(v), 'rows')
    if not v:
        continue
    fields = sorted({key for row in v for key in row.keys()})
    for row in v:
        vals = []
        for f in fields:
            val = row.get(f)
            if val is None:
                vals.append('NULL')
            elif isinstance(val, bool):
                vals.append('1' if val else '0')
            elif isinstance(val, (int, float)):
                vals.append(str(val))
            else:
                vals.append("'" + str(val).replace("'", "''") + "'")
        print('INSERT INTO `%s` (%s) VALUES (%s);' % (k, ','.join('`'+f+'`' for f in fields), ','.join(vals)))
